test/election.test.ts

```diff
@@ -0,0 +1,158 @@
+
+      await waitElection2;
+
+      await tearDownTestClient(client2);
+    });
+
+    it('should proclaim if campaign repeatly', async () => {
+      expect(election.isCampaigning).to.be.true;
+
+      const oldValue = await client.get(election.leaderKey);
+      expect(oldValue).to.equal('candidate');
+
+      await election.campaign('new-value');
+      const newValue = await client.get(election.leaderKey);
+      expect(newValue).to.equal('new-value');
+    });
+
+    it('only proclaim if value changed', async () => {
+      const proclaimFn = sinon
+        .stub(election, 'proclaim')
+        .callsFake(Election.prototype.proclaim.bind(election));
+
+      await election.campaign('candidate');
+      expect(proclaimFn.notCalled).to.be.true;
+      await election.campaign('candidate2');
+      expect(proclaimFn.calledOnce).to.be.true;
+
+      proclaimFn.restore();
+    });
+
+  });
+
+  describe('proclaim', () => {
+
+    it('should update if is a leader', async () => {
+      const oldValue = await client.get(election.leaderKey);
+      expect(oldValue).to.equal('candidate');
+      await election.proclaim('new-candidate');
+      const newValue = await client.get(election.leaderKey);
+      expect(newValue).to.equal('new-candidate');
+    });
+
+    it('should throw if not a leader', async () => {
+
+      await election.resign();
+      const whenCatch = sinon.spy();
+      await election.proclaim('new-candidate').catch(whenCatch);
+      expect(whenCatch.calledOnce).to.be.true;
+    });
+
+  });
+
+  describe('getLeader', () => {
+
+    it('should return leader key', async () => {
+      const leaderKey = await election.getLeader();
+      expect(election.leaderKey).to.equal(leaderKey);
+    });
+
+    it('should throw if no leader', async () => {
+      await election.resign();
+      const whenCatch = sinon.spy();
+      await election.getLeader().catch(whenCatch);
+      expect(whenCatch.calledOnce).to.be.true;
+    });
+
+  });
+
+  describe('observe', () => {
+    it('should emit leader event', async () => {
+      const client2 = new Etcd3(getOptions());
+      const election2 = new Election(client2, 'test-election');
+
+      let currentLeaderKey = '';
+      election.on('leader', leaderKey => currentLeaderKey = leaderKey);
+
+      expect(election.isObserving).to.be.true;
+
+      // looking for current leader and emit it
+      await sleep(30);
+
+      expect(currentLeaderKey).to.equal(election.leaderKey);
+
+      const waitElection2 = election2.campaign('candidate2');
+
+      await election.resign();
+
+      await waitElection2;
+
+      // waiting for watcher
+      await sleep(30);
+
+      expect(currentLeaderKey).to.equal(election2.leaderKey);
+    });
+
+    it('should wait for leader', async () => {
+      await election.resign();
+
+      let currentLeaderKey = '';
+      election.on('leader', leaderKey => currentLeaderKey = leaderKey);
+
+      // waiting for watcher created
+      await sleep(30);
```

Comment:
We will create a watcher to wait for new leader when no leader found. Seems if there is a new leader elected before watcher created, we will miss this leader. Is it possible?
If it is, maybe we can create watcher anyway before looking for current leader for avoid this case, although it will waste resource.